### PR TITLE
Use RefCountingRunnable when creating indices in TransportBulkAction

### PR DESCRIPTION
**Pull request by gmarouli**
When an index can be auto-created, the `TransportBulkAction` will first try to create all the missing indices and then it's going to index the documents. This means that it's going to perform a number of async create index requests. This number is equal to the number of distinct indices that can be auto-created in the request.

In this refactoring we use `RefCountingRunnable` to facilitate these async request. The `RefCountingRunnable` is already handling gracefully the different calls and can execute the bulk request when they all complete.

This will also makes the code easier to extend when adding more pre-requisite actions, such as https://github.com/elastic/elasticsearch/pull/103309.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 1]
```diff
@@ -390,52 +391,35 @@ protected void createMissingIndicesAndIndexData(
...
+            }
+        });
+        try (RefCountingRunnable refs = new RefCountingRunnable(executeBulkRunnable)) {
             for (String index : autoCreateIndices) {
-                createIndex(index, bulkRequest.timeout(), new ActionListener<>() {
-                    @Override
-                    public void onResponse(CreateIndexResponse result) {
-                        if (counter.decrementAndGet() == 0) {
-                            forkExecuteBulk(listener);
+                createIndex(index, bulkRequest.timeout(), refs.acquireListener().delegateResponse((l, e) -> {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:407]: Chaining `delegateResponse` and `map` in one expression is kind of suspicious, you're effectively overriding both `onResponse` and `onFailure` at the same time. Maybe just create a `new ActionListener` directly, with something like `ActionListener.releaseAfter(new ActionListener<>{...}, refs.acquire())`.

**Review by DaveCTurner — COMMENTED** [review]
Looks ok to me; I left a suggestion.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java (hunk 1 above)
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:407]: Thank you @DaveCTurner ! Good point, let me try it.

**Comment by gmarouli** [comment]
I see some failures locally during manual testing. I am investigating further.

**Comment by gmarouli** [comment]
@elasticmachine update branch

**Comment by gmarouli** [comment]
> I see some failures locally during manual testing. I am investigating further.

After extensive testing I wasn't able to reproduce it. What I saw initially, was Kibana reporting that the index already exists. 
I haven't been able to reproduce it since so I am considering maybe my local set-up was a bit off and after the clean rebuild it was fixed.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by gmarouli** [comment]
@DaveCTurner thank you for the feedback and the quick reactions!

**Pull request merged**